Bluetooth: host: Initial code for 5.2 features

Vudentz opened the pull request:
This add support for for the following features:

- L2CAP: Add support for Enhanced Credit Based Flow Control
- ATT/GATT: Add Enhanced ATT support
- ATT/GATT: Add Notify Multiple and Read Multiple Variable Length procedures

Vudentz committed:
Bluetooth: L2CAP: Add definitions introduced in 5.2

This introduces new Enhanced Credit Based Flow Control PDUs and related
definitions which were introduced in 5.2.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Add initial implementation of ECRED mode

This adds the initial implementation of ECRED mode which can connect up
to 5 channels simultaneously and is required by EATT.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Add definitions from 5.2

This adds the definitions for Enhanced ATT along with new PDUs and UUIDs
introduced in 5.2.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Add released callback

This adds a callback to indicate when the stack has released all
references to a given channel so the owner free up any resources
associated with that.

This is requires since EATT channels cannot rely on the destroy callback
as it does not use a fixed channel like ATT.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Add support EATT bearer

This adds support for EATT bearer which was introduced in 5.2, they work
as extra channels to have GATT traffic, at the moment it is completely
transparent to application when they are in use since the allocation
happens automatically.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add support for new PDUs

This adds support for ATT_MULTIPLE_HANDLE_VALUE_NTF,
ATT_READ_MULTIPLE_VARIABLE_REQ and ATT_READ_MULTIPLE_VARIABLE_RSP.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Process channel with fixed PSM on RX thread

This leaves only channels on the dynamic range to be offloaded to the
system queue so ATT and EATT handling are handling in the same context.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -451,15 +461,51 @@ static int l2cap_le_conn_req(struct bt_l2cap_le_chan *ch)
+			req->mtu = sys_cpu_to_le16(ch->rx.mtu);
+			req->mps = sys_cpu_to_le16(ch->rx.mps);
+			req->credits = sys_cpu_to_le16(ch->rx.init_credits);
+		}
```
You could consider moving this before the for-loop so you don't need to have a branch

Vudentz replied:
Will do.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -784,7 +842,7 @@ static void l2cap_chan_tx_give_credits(struct bt_l2cap_le_chan *chan,
 	atomic_add(&chan->tx.credits, credits);
 
-	if (!atomic_test_and_set_bit(chan->chan.status, BT_L2CAP_STATUS_OUT) &&
+	if (atomic_test_and_set_bit(chan->chan.status, BT_L2CAP_STATUS_OUT) &&
```
This doesn't look right

Vudentz replied:
Will fix this as well.

jhedberg commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
+	  Level 1 (BT_SECURITY_LOW) = No encryption or authentication required
+	  Level 2 (BT_SECURITY_MEDIUM) = Only encryption required
+	  Level 3 (BT_SECURITY_HIGH) = Encryption and authentication required
+	  Level 4 (BT_SECURITY_FIPS) = Secure connection required
```
We deprecated the LOW/MEDIUM/HIGH/FIPS defines, so I don't think they should be mentioned here (they're only called L1, L2, etc.

Vudentz replied:
Right, will update it.

aescolar commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/prj.conf:
```diff
@@ -24,8 +24,10 @@ CONFIG_BT_L2CAP_RX_MTU=512
 CONFIG_BT_L2CAP_TX_MTU=512
 CONFIG_BT_CTLR_RX_BUFFERS=3
 CONFIG_BT_DEBUG_LOG=y
+CONFIG_BT_EATT=n
```
CC: @heri-developer

Vudentz replied:
Just a remimder here, test shall not assume handle positions are static on the database, if a handle is required for a test perform a lookup into the database to resolve the handle at runtime.

wopu-ot replied:
Could you maybe create an issue for EDTT (https://github.com/EDTTool/EDTT) to describe what is broken/needs enhancement in EDTT when enabling EATT?

Vudentz replied:
I though this was quite clear by now, anyway autopts might needs adjusting as well.

heri-developer replied:
The Bluetooth-SIG GATT test database(s) does use static assigned handles, that is an essential part of the tests. Until they decide to change test-strategy, we need to be able to support that.

wopu-ot replied:
Comments in this PR will be forgotten as soon as it is merged. An issue will remind us what needs fixing in the future.

Vudentz replied:
> The Bluetooth-SIG GATT test database(s) does use static assigned handles, that is an essential part of the tests. Until they decide to change test-strategy, we need to be able to support that.

I guess you are refering to the fact the you enter the handles in pixit or something like that, but we don't have to follow how PTS does things here since EDTT just emulates the tests we can actually inform the handles at runtime, so note that the purpose of these tests are to tests the procedures not the actual database layout since GATT don't actually have a concept of fixed/pinned services, the order is undefined.

Vudentz replied:
https://github.com/EDTTool/EDTT/issues/16

heri-developer replied:
No - I'm referring to the official GATT Qualification Test DB that can be found here: https://www.bluetooth.org/docman/handlers/DownloadDoc.ashx?doc_id=229720.

jhedberg replied:
@heri-developer isn't selecting a SIG-provided database optional in the GATT IXIT, i.e. it's still possible to qualify stacks without such a database? I'm looking e.g. at GATT:P4.1 and GATT:P4.2 and both have a comment "Alternatively, a SIG-provided database may be used."

Vudentz replied:
Ive actually end up dropping this since EATT is no longer enabled by default.

joerchan reviewed (commented):
Sorry for taking so long to post feedback.
I haven't looked too closely at all of the implementation yet, so posting the more shallow comments. Will continue to look.

joerchan commented:
Regarding the RFC tag. Is that for the API part? If so I think that part looks good. But it would be nice to get feedback from users who are expected to use this API as well.

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -864,6 +864,17 @@ struct bt_gatt_notify_params {
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_notify_multiple(struct bt_conn *conn, u16_t num_params,
```
The _cb was added later to bt_gatt_notify and I don't think we would need 2 version here so I went with the more conventional form of the 2.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -631,8 +754,8 @@ static u8_t find_type_cb(const struct bt_gatt_attr *attr, void *user_data)
 
-	/* Fast forward to next item position */
-	data->group = net_buf_add(data->buf, sizeof(*data->group));
+	/* Fast foward to next item position */
```
Will fix it.

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -795,7 +868,7 @@ static void l2cap_chan_rx_give_credits(struct bt_l2cap_le_chan *chan,
 	BT_DBG("chan %p credits %u", chan, credits);
 
-	atomic_add(&chan->rx.credits, credits);
+	atomic_sub(&chan->rx.credits, credits);
```
Weird, I thought I had fixed this one, anyway it is wront to sub here.

Vudentz replied:
Fixed.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -716,6 +726,17 @@ static void remove_cf_cfg(struct bt_conn *conn)
+static ssize_t sf_read(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+		       void *buf, u16_t len, u16_t offset)
+{
+	u8_t value = BIT(0);
```
Sure I can add defines for them.

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -864,6 +864,17 @@ struct bt_gatt_notify_params {
 int bt_gatt_notify_cb(struct bt_conn *conn,
```
I was wondering about that at some point, but I guess the application can take the decision to group or not, besides Id like to avoid extra lockups with notification since that could be time sensitive if the user wants to notify as fast as possible.

Vudentz commented:
> Regarding the RFC tag. Is that for the API part? If so I think that part looks good. But it would be nice to get feedback from users who are expected to use this API as well.

I can drop the RFC tag then, I was hoping to get more feedback but it looks like either there were no comments or people were busy with something else.

joerchan commented:
> I can drop the RFC tag then, I was hoping to get more feedback but it looks like either there were no comments or people were busy with something else.

@aescolar @Rasmus-Abildgren-Bose @carlescufi Any comments on the proposed API changes?

Rasmus-Abildgren-Bose commented:
I'm currently too busy to do a full review but have had a look at the API.

I see no problem with these changes.

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -451,15 +461,57 @@ static int l2cap_le_conn_req(struct bt_l2cap_le_chan *ch)
-	if (chan->state != BT_L2CAP_CONNECT || chan->ident) {
+	/* Skip channels already connected or there is an ongoing request */
+	if (chan->state != BT_L2CAP_CONNECT ||
+	    k_delayed_work_remaining_get(&chan->rtx_work)) {
```
Ive added a check for k_work_pending as well it the timer may have expired but the work has not been executed yet, note though that this is not canceling the work it is just checking if there a pending request which means the channel is not waiting for a encryption change to retry connecting.

joerchan replied:
I think that could still be problematic. Since the timeout might have been expired, but not yet handled.
I tried to create helper function to address this issue, which should handle the case where the remaining value is zero. But @pabigot still didn't think this was safe enough:
https://github.com/zephyrproject-rtos/zephyr/pull/22826/files#diff-ed0b4d3815d3ce00b451fa30ed3c2051R143

In the Mesh similar problem is solved through state-flags instead.

pabigot replied:
A better summary of my position is https://github.com/zephyrproject-rtos/zephyr/pull/22826#pullrequestreview-371142029.  I originally blocked that PR due to lack of a test; it seems to have a test, so I've removed my NAK.  Until there's a more robust `k_work` implementation I also an using external state to track at a gross level whether a submitted request is still pending.

Vudentz replied:
That sounds a little too paranoid here given that there is a state already, but I will go ahead and create a flag as well.

Vudentz committed:
Bluetooth: L2CAP: Use spaces to align defines

This replaces the use of tabs with spaces to align defines.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented on subsys/bluetooth/host/l2cap_internal.h:
```diff
@@ -190,6 +191,40 @@ struct bt_l2cap_le_credits {
+	u16_t scid[0];
+} __packed;
+
+#define BT_L2CAP_RECONF_SUCCESS		0x0000
```
I added a patch replacing tabs with spaces to align the defines.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -733,20 +854,98 @@ static u8_t err_to_att(int err)
+		if (read < 0) {
+			if (total) {
+				return total;
+			}
```
Fixed.

carlescufi reviewed (approved):
No objections to the API additions (I did not see any actual changes, only additions)

Vudentz commented:
> No objections to the API additions (I did not see any actual changes, only additions)

Yep, it is mostly transparent from application point of view, there are some additions to enable new procedures such as notify multiple but other than that application will not need to do anything in order to use these features.

jhedberg commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
+	  This option enables support for Enhanced ATT bearers support. When
+	  enabled additional L2CAP channels can be connected as bearers enabling
+	  multiple outstanding request.
+
```
Agreed. Any time you have two or more Kconfig options depending on the same thing (or anticipate it in the near future) `if <foo>` is preferable.

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/Kconfig.l2cap:
```diff
@@ -50,6 +52,7 @@ config BT_L2CAP_TX_MTU
 
 config BT_L2CAP_DYNAMIC_CHANNEL
 	bool "L2CAP Dynamic Channel support"
+	default y
```
How come?

jhedberg replied:
I was already surprised that EATT defaulted to `y` but thought it's not that bad since it depends on DYNAMIC_CHANNEL, but I now see you default that also to `y` which pulls everything in unless explicitly disabled, which doesn't seem wise.

Vudentz replied:
Right, we can have EATT select BT_L2CAP_DYNAMIC_CHANNEL instead of defaulting to y.

Vudentz replied:
Looks like we cannot really use select:

```
warning: BT_L2CAP_DYNAMIC_CHANNEL (defined at subsys/bluetooth/host/Kconfig.l2cap:53) has direct dependencies BT_SMP && BT_CONN && BT_HCI_HOST && BT_HCI && BT with value n, but is currently being y-selected by the following symbols:
 - BT_EATT (defined at subsys/bluetooth/host/Kconfig.gatt:37), with value y, direct dependencies BT_CONN && BT_HCI_HOST && BT_HCI && BT (value: y), and select condition BT_CONN && BT_HCI_HOST && BT_HCI && BT (value: y)
```

The other option is left EATT as optional, though I guess later on we would like to have EATT enabled by default if BT_L2CAP_DYNAMIC_CHANNEL is enabled, so perhaps we can use default y if BT_L2CAP_DYNAMIC_CHANNEL.

Vudentz replied:
So I resolve this by having a dependency to BT_L2CAP_DYNAMIC_CHANNEL in addition to default y if BT_L2CAP_DYNAMIC_CHANNEL.

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -1892,6 +2207,10 @@ static const struct att_handler {
 		sizeof(struct bt_att_indicate),
 		ATT_INDICATION,
 		att_indicate },
+	{ BT_ATT_OP_NOTIFY_MULT,
+		sizeof(struct bt_att_notify_mult),
+		ATT_NOTIFICATION,
+		att_notify_mult },
```
This is not an EATT operation? (no #ifdef around it)

Vudentz replied:
No, the new procedures are not exclusive to EATT bearer, they just need to be enable by the client but can be used on ATT bearer as well.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3305,10 +3515,81 @@ static int gatt_read_multiple(struct bt_conn *conn,
+		return;
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)pdu, length);
```
Maybe add a code comment above this since otherwise it looks quite dangerous (casting `const void *` to `void *`). E.g. mention that it's safe to do since none of the `net_buf_simple` operations you do on the buffers modifies the data in it.

Vudentz replied:
Will do.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -903,63 +1038,200 @@ static void le_conn_req(struct bt_l2cap *l2cap, u8_t ident,
+
+	BT_DBG("mtu %u mps %u", mtu, mps);
+
+	result = BT_L2CAP_RECONF_SUCCESS;
```
You might want to move the this after the loop, since you're needlessly writing twice to that variable in the invalid MTU case.

Vudentz replied:
Done.

wopu-ot commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/prj.conf:
```diff
@@ -24,8 +24,10 @@ CONFIG_BT_L2CAP_RX_MTU=512
 #
 CONFIG_BT_DEVICE_NAME_DYNAMIC=y
-CONFIG_LOG_MAX_LEVEL=2
+CONFIG_LOG_MAX_LEVEL=4
```
Why change the default log level?

Vudentz replied:
This was for debugging, there is very little information printed with level 2 not sure why is was done like that.

Vudentz replied:
Fixed.

Vudentz commented:
@jhedberg All comments have been addressed, with the latest changes EATT won't be enable by default anymore as it depends on L2CAP CoC, so I guess it is safe to be merged.

jhedberg commented:
@Vudentz there seems to be several unanswered comments from @joerchan

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -864,6 +864,17 @@ struct bt_gatt_notify_params {
 
+/** @brief Notify multiple attribute value change.
+ *
+ *  @param conn Connection object.
```
If I were to update all the documentation with this sort of comment that will take a while, so how about we do this when we decided its the right thing to do for the entire set of APIs?

Vudentz commented on include/bluetooth/l2cap.h:
```diff
@@ -332,6 +339,21 @@ int bt_l2cap_server_register(struct bt_l2cap_server *server);
+ *  rejected disconnected() callback is called instead.
+ *
+ *  @param conn Connection object.
+ *  @param chans Channel objects
```
Added the array into the documentation.

Vudentz commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
 
+config BT_EATT
+	bool "Enhanced ATT Bearers support"
+	default y
```
Well I guess we want to promote the use of EATT since it does have benefits over regular ATT and the footprint is really minimal when BT_L2CAP_DYNAMIC_CHANNEL is already enabled, it is kind weird to see this sort of comment comming from you btw, since default n tends to create a lot more problem down the road, for instance if I had done this since the beginning EDTT sample would be broken given it wouldn't even test EATT.

joerchan replied:
I think any feature addition should be default to `n` and then we can enable the feature for the tests.

Vudentz replied:
Are you saying that because it is experimental it should default to n? There are already testing for it since all tests running GATT tests will essentially enable it, anyway it is already gated by BT_L2CAP_DYNAMIC_CHANNEL, that said maybe we can wait to enable it when qualification is done but that usually is a past a release so that means no new feature can default to y until qualified.

Vudentz replied:
I marked it as EXPERIMENTAL like it was done for Advertising Extensions.

Vudentz commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
+config BT_EATT_RX_MTU
+	int "Maximum supported Enhanced ATT MTU for incoming data"
+	default 70
+	range 70 519
```
It is to include L2CAP header + ATT opcode up to 512 bytes which is the limit of a GATT attribute, after that it becomes unusable for the likes of regular read and writes, so in theory only discover and the likes of read by type can really make use of more than 512 bytes.

joerchan replied:
Ok. It might be worth to put that explanation somewhere, for example like it is done for `BT_RX_BUF_LEN`

Vudentz replied:
Done.

Vudentz commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
+	range 1 4
+	depends on BT_EATT
+	help
+	  Security level of EATT bearers:
```
I can add that but nowhere in the APIs we express the concept of security level for client APIs.

Vudentz commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -34,6 +34,53 @@ config BT_ATT_TX_MAX
+config BT_GATT_NOTIFY_MULTIPLE
+	bool "GATT Notify Multiple Characteristic Values support"
+	depends on BT_EATT
+	depends on BT_GATT_CACHING
```
Well take will lead to a very complicate design where certain bits would be supported and others don't, which imo is not worth the trouble, that said notify multiple does not actually depend on EATT, any bearer can use it so I will fix it.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -105,122 +125,192 @@ static void att_req_destroy(struct bt_att_req *req)
+			return -ENOTSUP;
+		}
+
+		if (!atomic_test_bit(chan->chan.chan.status,
+				     BT_L2CAP_STATUS_OUT)) {
+			return -EAGAIN;
+		}
```
bt_l2cap_chan_send will just queue the buf thus why the check is here so we continue attempting with another channel which may be available immediatelly, in case none can send the buf immediatelly it is put back to ATT queue.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -275,129 +372,140 @@ static u8_t att_mtu_req(struct bt_att *att, struct net_buf *buf)
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&att->chans, chan, tmp, node) {
+		/* If there is nothing pending use the channel */
+		if (!chan->req) {
+			if (bt_att_chan_req_send(chan, ATT_REQ(node)) >= 0) {
```
That means it was able to send since the channel can be out of credits it may not be able to send the request immediatelly.

joerchan replied:
Right. `bt_att_chan_req_send` returns number of bytes as an int. That's what I missed. Thanks.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -593,10 +701,25 @@ static u8_t find_type_cb(const struct bt_gatt_attr *attr, void *user_data)
+		if (!frag) {
+			return BT_GATT_ITER_STOP;
+		}
+		net_buf_frag_add(data->buf, frag);
```
Fixed.

Vudentz commented:
@joerchan patches have been updated to address your comments.

joerchan commented on include/bluetooth/l2cap.h:
```diff
@@ -257,6 +257,13 @@ struct bt_l2cap_chan_ops {
 	 */
 	void (*status)(struct bt_l2cap_chan *chan, atomic_t *status);
+
+	/* Channel released callback
```
Needs `@brief` here (I have a PR that fixes the rest that are missing).

Vudentz replied:
Done.

Vudentz committed:
Bluetooth: shell: Enable BT_EATT

This enables BT_EATT for shell samples so it can be build test by CI and
gives user the ability of test it using shell commands.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@joerchan Patches updates, EATT is not marked as experimental.

Vudentz committed:
Bluetooth: L2CAP: Add status flag to track encrypt pending

This adds a new flag to track if the L2CAP channel is pending waiting
for encryption to be changed to resume connecting.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi commented:
@joerchan can you please test this on a complex application?

Vudentz commented:
> @joerchan can you please test this on a complex application?

Preferable against something that implements EATT already like BlueZ, but you will need to update the enable the kernel parts if you want to test ecred procedures as otherwise it connects L2CAP channel by channel, that said the features is marked as experimental exactly to give people the opportunity to test this themselves.

joerchan commented:
@Vudentz My main concern was regression into ATT since some of the refactoring touches on common code. We have tested the nrf_desktop application in NCS with these changes to check for regressions. We haven't found any issues.

I don't have anything to test with for the new features, so I will just go over the latest changes.

Vudentz commented:
> @Vudentz My main concern was regression into ATT since some of the refactoring touches on common code. We have tested the nrf_desktop application in NCS with these changes to check for regressions. We haven't found any issues.
> 
> I don't have anything to test with for the new features, so I will just go over the latest changes.

I though the reason to add EDTT as part of CI was to extensively test ATT already, anyway now I have to resolve the conflicts you introduced with changes to the documentation which could have been avoided by just waiting this set to go in.

joerchan commented on subsys/bluetooth/host/att.c:
```diff
@@ -2291,33 +2690,40 @@ void bt_att_req_free(struct bt_att_req *req)
+		k_fifo_put(&att->tx_queue, buf);
 	}
 
 	return 0;
```
Should this now return `ret`?

Vudentz replied:
Nope, it is queueing if there is an error.

joerchan commented:
@Vudentz nrf_desktop is a HID mouse application. It has caught many regressions earlier which is why I wanted to test it now as well. EDTT test cases are good, but they test one thing at a time and are mostly functional test-cases. Some stress-testing is good as well.

This is looking good to me. I only have the one question.
